active-rfcs/0000-vuex-class-style-api.md

```diff
@@ -0,0 +1,440 @@
+the motivation could be used to develop alternative solutions. In other words,
+enumerate the constraints you are trying to solve without coupling them too
+closely to the solution you have in mind.
+-->
+
+The main motivation is type safety. The current object-style API is hard to infer
+state, getters, actions and mutations types in a module because they are circularly referenced by each other.
+
+There is [an attempt](https://github.com/vuejs/vuex/pull/1121) to type with the current API but it is tricky –
+It requires separated type declarations for module assets and manually annotating them to a module implementation.
+It could be also verbose as we need to write action name both on action type and its implementation, for example.
+
+The class-style API itself provides type-safety **in** modules themselves which means all usage for local state, getters,
+actions and mutations are strictly typed. However it is not typed that accessing a module asset from a component or
+another module yet. This will covered by another proposal that is using module type information which class-style syntax provides.
+https://github.com/vuejs/rfcs/pull/15
+
+The class-style API also solves the case that we need to manage some external object in actions which is not suitable
+to be in state. In the current object-style API, we need to put a variable to retain such object out of actions object
+but the variable is actually shared if the actions are reused on the multiple places in a store.
+
+```js
+// The current work around for managing an external object in actions.
+
+// Declare varable to retain WebSocket connection.
+// But this variable is a shared object.
+let ws
+
+const actions = {
+  init() {
+    ws = new WebSocket('wss://...')
+  },
+
+  emit() {
+    ws.send(/* ... */)
+  }
+}
+```
+
+# Detailed design
+
+<!--
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Vue to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+-->
+
+Vuex provides the following classes to create module:
+
+- `Getters`
+- `Mutations`
+- `Actions`
+
+They should be imported by ES import statement:
+
+```js
+import { Getters, Mutations, Actions } from 'vuex'
+```
+
+## Module state
+
+Module state can be defined in class-style syntax:
+
+```ts
+class CounterState {
+  count = 0
+}
+```
+
+By class-style syntax, the user defines both type and value of state object in one time.
+That means the `CounterState` in the above example can be passed as state value while
+used for other module assets as type.
+
+## Getters base class
+
+`Getters` class has following properties:
+
+- `state`: module local state.
+- `getters`: module local getters.
+- `rootState`: root state.
+- `rootGetters`: root getters.
```

Comment:
So technically we need three type parameters for `Getters` class, right? `rootState` and `rootGetters` are optional. `getters` should be typed via `this`.

Reply:
I intentionally exclude root things from type parameters because they make types too complicated. I also realized we do not need to use them when we use module context which is proposed on #15.

> `getters` should be typed via `this`.

Yeah right.